8313265: Move the FFM API out of preview

JornVernee opened the pull request:
Move the FFM API out of preview, by removing preview feature annotations from the implementation, and adjusting the tests to no longer use `--enable-preview`

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8313265](https://bugs.openjdk.org/browse/JDK-8313265): Move the FFM API out of preview (**Enhancement** - P2)


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - Committer)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign.git pull/853/head:pull/853` \
`$ git checkout pull/853`

Update a local copy of the PR: \
`$ git checkout pull/853` \
`$ git pull https://git.openjdk.org/panama-foreign.git pull/853/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 853`

View PR using the GUI difftool: \
`$ git pr show -t 853`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/853.diff">https://git.openjdk.org/panama-foreign/pull/853.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/panama-foreign/pull/853#issuecomment-1654582176)

JornVernee commented on src/java.base/share/classes/jdk/internal/javac/PreviewFeature.java:
```diff
@@ -64,10 +64,10 @@
+        // not used, but required for interim javac to not warn.
         VIRTUAL_THREADS,
-        @JEP(number=442, title="Foreign Function & Memory API", status="Third Preview")
         FOREIGN,
```
Note that I had to leave the enum constant here for now, since otherwise we encounter a warning when building interim javac.

```
jorn@Jorn-PC:/mnt/h/openjdk/foreign-abi$ make run-test-jdk_foreign CONF=windows-fastdebug
Using exact match for CONF=windows-fastdebug (other matches are possible)
Building target 'run-test-jdk_foreign' in configuration 'windows-fastdebug'
warning: unknown enum constant Feature.FOREIGN
error: warnings found and -Werror specified
1 error
1 warning
gmake[3]: *** [CompileInterimLangtools.gmk:127: /mnt/h/openjdk/foreign-abi/build/windows-fastdebug/buildtools/interim_langtools_modules/java.compiler.interim/_the.BUILD_java.compiler.interim_batch] Error 1
gmake[2]: *** [make/Main.gmk:78: interim-langtools] Error 2
gmake[2]: *** Waiting for unfinished jobs....

ERROR: Build failed for target 'run-test-jdk_foreign' in configuration 'windows-fastdebug' (exit code 2)

=== Output from failing command(s) repeated here ===
* For target buildtools_interim_langtools_modules_java.compiler.interim__the.BUILD_java.compiler.interim_batch:
warning: unknown enum constant Feature.FOREIGN
error: warnings found and -Werror specified
1 error
1 warning

* All command lines available in /mnt/h/openjdk/foreign-abi/build/windows-fastdebug/make-support/failure-logs.
=== End of repeated output ===

No indication of failed target found.
HELP: Try searching the build log for '] Error'.
HELP: Run 'make doctor' to diagnose build problems.

make[1]: *** [/mnt/h/openjdk/foreign-abi/make/Init.gmk:323: main] Error 2
make: *** [/mnt/h/openjdk/foreign-abi/make/Init.gmk:189: run-test-jdk_foreign] Error 2
```

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/AddressLayout.java:
```diff
@@ -55,7 +54,6 @@
  * @see #ADDRESS_UNALIGNED
  * @since 19
```
I seem to recall that `@since` should be adjusted to state the version in which we leave preview.

JornVernee replied:
Ok, I'll try to get confirmation on that, and then adjust the `@since` tags as needed.

JornVernee replied:
It seems that, at least this was also done in the case of Loom: https://github.com/openjdk/jdk/pull/13203/files#diff-2a5b4b651fa41e70f7953f7ee06e964d8396fcd1fe5755a1609e81b3f84d9224R842

JornVernee replied:
I've confirmed that we need to update the `@since` tags. This is also mentioned in [JEP12](https://bugs.openjdk.org/browse/JDK-8195734): 

> The author of a preview API element must also provide an `@since` tag in the element's javadoc comment, indicating the release when `@preview` was first added. If the API element is eventually made final and permanent in Java SE $N, then the `@since` tag must be changed to indicate release $N, as the element's history prior to $N is not of long-term interest.

mcimadamore reviewed (commented):
Looks good. I believe there's probably other tests to fix - surely Vector tests, and also a couple in stream/spliterator and other areas?

JornVernee commented:
> Looks good. I believe there's probably other tests to fix - surely Vector tests, and also a couple in stream/spliterator and other areas?

Yes, good point. I will check those as well.

JornVernee commented:
- I've updated the `@since` tags
- Updated the benchmarks, which I had forgotten before
- Searched the test directory for java.lang.foreign and updated all tests that used the API.

mcimadamore reviewed (approved):
Looks good. Re. benchmarks, I seem to recall that we also added --enable-preview in the command to compile them. I presume we want to keep the makefile as is, for other preview APIs?

JornVernee commented:
> Looks good. Re. benchmarks, I seem to recall that we also added --enable-preview in the command to compile them. I presume we want to keep the makefile as is, for other preview APIs?

Yes, there are ScopedValue and String Template benchmarks that depend on that.

JornVernee commented:
/integrate
